lib/minurl.shared.js

```diff
@@ -0,0 +1,32 @@
+/**
+ * @typedef URL
+ * @property {string} hash
+ * @property {string} host
+ * @property {string} hostname
+ * @property {string} href
+ * @property {string} origin
+ * @property {string} password
+ * @property {string} pathname
+ * @property {string} port
+ * @property {string} protocol
+ * @property {string} search
+ * @property {any} searchParams
+ * @property {string} username
+ * @property {() => string} toString
+ * @property {() => string} toJSON
+ */
+
+/**
+ * @param {unknown} fileURLOrPath
+ * @returns {fileURLOrPath is URL}
+ */
+export function isUrl(fileURLOrPath) {
+  return (
+    fileURLOrPath !== null &&
+    typeof fileURLOrPath === 'object' &&
+    // @ts-expect-error: indexable.
+    fileURLOrPath.href &&
+    // @ts-expect-error: indexable.
+    fileURLOrPath.origin
+  )
```

Comment:
This could just be `fileURLOrPath instanceof URL`, but the types are an issue here.

What about creating the following file:

```js
export { URL };
```

With the following type definition, which must be hand written, not generated:

```ts
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34960#issuecomment-576906058
export const URL: typeof globalThis extends { URL: infer URLCtor }
  ? URLCtor
  : typeof import('url').URL;
```

Then you could import `URL` from this file.

---

I don’t like this solution myself, but it might work.

Reply:
a) I’m personally not a fan of `instanceof`, b) this was ported from https://github.com/nodejs/node/blob/fcf8ba4269eb4b9984f9f0d926e0d8e0bd6d4382/lib/internal/url.js#L1501-L1503 (but I didn‘t make that clear)

---

Per my testing, that `URL` can’t be exported. This is also noted in the follow-up comment to what you link: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34960#issuecomment-576922826.

I tried being smart about that, I completely failed 🤷‍♂️

Reply:
The following works in both Node and the browser, despite TypeScript throwing type errors:

```ts
// @ts-expect-error
return fileURLOrPath instanceof URL
```

Reply:
added a note on where that function is ported from

Reply:
One part is problems with multiple versions. I recently hit that a lot when updating all projects, different versions of unified / vfile.
Other than packages from npm it also arises in multiple frames/windows.
While it’s a global, I can imagine (not sure) bundlers still adding it for some reason, and perhaps adding different ones in different places?

But also more philosophically that I have a good experience in JS with APIs that check for *looks like a thing* rather than *is a subclass of thing* — I feel being relaxes about input is in line with the good (and bad) properties of the language